### PR TITLE
implemented animalUtil class

### PR DESCRIPTION
**Pull request by Kuahoo**

### CODE REVIEW

File: AnimalTracker/app/src/main/java/toombs/animaltracker/DummyPage.java [hunk 1]
```diff
@@ -8,25 +11,64 @@
...
         setContentView(R.layout.activity_dummy_page);
         Toolbar toolbar = findViewById(R.id.toolbar);
         setSupportActionBar(toolbar);
 
+        //My very ugly testing of the animal class
+        prefs = getSharedPreferences("animalPreferences", MODE_PRIVATE); // creates animalPreferences file and has parameter firstRun to check if app has ran before
+        if(prefs.getBoolean("firstRun", false)){
+        }else{
+            HashSet<String> animalListSet = new HashSet<String>();
+            Animal A = new Animal("Dingus", "N/A", "Zeus", new GregorianCalendar(2020, 9, 5, 5, 5), "M", true, null);
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/DummyPage.java:54]: OMEGALUL

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 2]
```diff
@@ -1,4 +1,107 @@
...
+            return out.toByteArray();
+
+        }catch(IOException e){
+            Log.i("Exception Error:", e.getMessage());
+            return null;
+        }
+    }
+
+
+    public static String createFileName(Animal animal) {
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:102]: The fileName for the animal should just be its hash.

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 3]
```diff
@@ -1,4 +1,107 @@
...
+            while(hashtext.length() < 32){
+                hashtext = "0" + hashtext;
+            }
+            return hashtext;
+        }catch (NoSuchAlgorithmException e){
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static byte[] ObjectToByteArray(Object obj){
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:88]: this method should only be used in this file, private.

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 4]
```diff
@@ -1,4 +1,107 @@
...
+
+    public static String hashObject(byte[] byteArray){
+
+        try{
+            MessageDigest md = MessageDigest.getInstance("SHA-1");
+            byte[] messageDigest = md.digest(byteArray);
+            BigInteger no = new BigInteger(1, messageDigest);
+            String hashtext = no.toString(16);
+            while(hashtext.length() < 32){
+                hashtext = "0" + hashtext;
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:80]: :) this should be done with strbuffer or something else.

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 5]
```diff
@@ -1,4 +1,107 @@
...
+            ois.close();
+
+            return (HashSet<String>) deserializeAnimalSet;
+        }catch(IOException | ClassNotFoundException e){
+            Log.i("Exception error:", Objects.requireNonNull(e.getMessage()));
+            return null;
+        }
+    }
+
+    public static String hashObject(byte[] byteArray){
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:72]: this should probably also be made private. There should be a method "serializeAnimal" that calls serialize object, but performs some extra steps that serialize object doesn't perform (generates hash if not already present, checks that hash is present in set, puts in set if not and reserializes set, then serializes itself).

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 6]
```diff
@@ -1,4 +1,107 @@
...
+            ObjectOutputStream oos = new ObjectOutputStream(fos);
+            oos.writeObject(obj);
+            fos.close();
+        }catch (IOException e){
+            Log.i("Exception error:", Objects.requireNonNull(e.getMessage()));
+        }
+    }
+
+
+    public static Animal deserializeAnimal(Context context, String fileName){
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:36]: DRY

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 7]
```diff
@@ -1,4 +1,107 @@
...
+            Log.i("Exception error:", Objects.requireNonNull(e.getMessage()));
+            return null;
+        }
+
+    }
+
+
+    //Suppressed warning unchecked because Java doesn't like me returning an unchecked HashSet. We need to check if it's a String hashset.
+    @SuppressWarnings(value = "unchecked")
+    public static HashSet<String> deserializeAnimalSet(Context context){
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:54]: DRY

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java [hunk 8]
```diff
@@ -1,4 +1,107 @@
...
+import java.math.BigInteger;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashSet;
+import java.util.Objects;
+
 public class AnimalUtil {
+    public static String animalFilePath = "/files/AnimalData/";
+    public static String animalSetFilePath = "/files/AnimalData/animalSet.ser";
+    public static void serializeObject(Context context, Object obj, String fileName){
```
Comment (Timothy-Toombs) [inline AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:23]: One of two strategies can be adopted here:
1. make this method private and have  serializeSet and serializeAnimal methods that call this by proxy
2. make this method public but only call it in other classes when serializing the set. (making serializeAnimal function is necessary, see below).

**Review by Timothy-Toombs — CHANGES_REQUESTED** [review]

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java (hunk 6 above)
Comment (chrissclipei) [reply AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:36]: WET

File: AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java (hunk 7 above)
Comment (chrissclipei) [reply AnimalTracker/app/src/main/java/toombs/animaltracker/animals/AnimalUtil.java:54]: WET

**Review by chrissclipei — COMMENTED** [review]
Don't listen to Tim Kaho he's just a big bully YOU GOT THIS!!! SHOW HIM HOW TO CODE!!!

**Pull request closed without merging**